packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -49,50 +57,76 @@ class ProtocolDiscovery {
   final int devicePort;
   final bool ipv6;
 
-  final Completer<Uri> _completer = Completer<Uri>();
+  /// The time to wait before forwarding observatory URIs from [logReader].
+  final int throttleTimeInMilliseconds;
 
   StreamSubscription<String> _deviceLogSubscription;
 
+  StreamController<Uri> _uriStreamController;
+
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri async {
+    try {
+      return await uris.first;
+    } on StateError {
+      // If the stream is closed and empty.
+      return null;
+    }
+  }
+
+  /// The discovered service URI stream.
   ///
   /// Port forwarding is only attempted when this is invoked, in case we never
   /// need to port forward.
-  Future<Uri> get uri async {
-    final Uri rawUri = await _completer.future;
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        timeInMilliseconds: throttleTimeInMilliseconds,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) => _forwardPort(observatoryUri));
   }
 
   Future<void> cancel() => _stopScrapingLogs();
 
   Future<void> _stopScrapingLogs() async {
+    await _uriStreamController?.close();
     await _deviceLogSubscription?.cancel();
     _deviceLogSubscription = null;
   }
 
-  void _handleLine(String line) {
-    Uri uri;
+  Match _getPatternMatch(String line) {
     final RegExp r = RegExp('${RegExp.escape(serviceName)} listening on ((http|\/\/)[a-zA-Z0-9:/=_\\-\.\\[\\]]+)');
-    final Match match = r.firstMatch(line);
+    return r.firstMatch(line);
+  }
 
+  Uri _getObservatoryUri(String line) {
+    final Match match = _getPatternMatch(line);
     if (match != null) {
-      try {
-        uri = Uri.parse(match[1]);
-      } on FormatException catch (error, stackTrace) {
-        _stopScrapingLogs();
-        _completer.completeError(error, stackTrace);
-      }
+      return Uri.parse(match[1]);
+    }
+    return null;
+  }
+
+  void _handleLine(String line) {
+    Uri uri;
+    try {
+      uri = _getObservatoryUri(line);
+    } on FormatException catch(error, stackTrace) {
+      _uriStreamController.addError(error, stackTrace);
```

Comment:
Note that if the listener on `_uriStreamController.stream` doesn't have an `onError` callback, then this error will be thrown into the containing zone.

Reply:
I added the `onError` callback. Thanks.